src/compiler/parser.ts

```diff
@@ -1001,10 +1008,12 @@ module ts {
             switch (parsingContext) {
                 case ParsingContext.SourceElements:
                 case ParsingContext.ModuleElements:
-                    return isSourceElement(inErrorRecovery);
+                    // During error recovery we don't treat empty statements as statements
+                    return !(token === SyntaxKind.SemicolonToken && inErrorRecovery) && isModuleElement();
                 case ParsingContext.BlockStatements:
                 case ParsingContext.SwitchClauseStatements:
-                    return isStartOfStatement(inErrorRecovery);
+                    // During error recovery we don't treat empty statements as statements
+                    return !(token === SyntaxKind.SemicolonToken && inErrorRecovery) && isStatement();

```

Comment:
I've given this a lot of thought, and i'm very opposed to what's happened here.  Originally the code was very clear in intent and placed responsibility in the right location.  This list parsing code simply asked:

> does it look like we're starting a statement here, depending on whether we're in error recovery or not?

That function that was called did the right thing, and could be nicely composed in other function (for example, in isSourceElement).  Now, you've taken out the existing error recovery case this used to handle (semicolons during error recovery) and inlined that logic in 3 separate places instead of leaving it directly at the point where we see a semicolon and need to determine if it is a valid statement or not.

This is non-ideal for two reasons:

1.  It violates DRY (don't repeat yourself).  Instead of having a single place where we deal with error recovery and statements, we now have three places.
2. It means any further error recovery work we do around statements will have to be put in a multiplicity of places that the author will need to discover and hunt down.  As error recovery improvements are always happening throughout the life of a language, i would expect us to have to do work here in the future.  The new approach makes this much more difficult.

The original code expresses intent better, and provides a better solution for code composition.  I would prefer we move back that style.  Looking at the structure of hte new code, i don't see any reason at all why passing along inErrorRecovery would be a problem.